### PR TITLE
Make ITab an unsealed runtimeclass

### LINKED ISSUE #7969: Close Menu, "Close other tabs" and "Close tabs to right" always operate on tab index 0, are always enabled
I'm not sure if this is an issue in main or an issue in feature/settings-ui.

/cc @mpela81

### PR DESCRIPTION
**Pull request by carlos-zamora**
`TerminalTab` and `SettingsTab` share some implementation details. The
close submenu introduced in #7728 is a good example of functionality
that is consistent across all tabs. This PR transforms `ITab` from an
interface, into an [unsealed runtime class] to de-duplicate some
functionality. Most of the logic from `SettingsTab` was moved there
because I expect the default behavior of a tab to resemble the
`SettingsTab` over a `TerminalTab`.

## References
Verified that Close submenu work was transferred over (#7728, #7961, #8010).

## Validation Steps Performed
Check close submenu on first/last tab when multiple tabs are open.

Closes #7969

[unsealed runtime class]: https://docs.microsoft.com/en-us/uwp/midl-3/intro#base-classes

### CODE REVIEW

File: src/cascadia/TerminalApp/TerminalTab.cpp [hunk 1]
```diff
@@ -189,8 +189,12 @@ namespace winrt::TerminalApp::implementation
             Icon(_lastIconPath);
             TabViewItem().IconSource(IconPathConverter::IconSourceMUX(_lastIconPath));
 
-            // Update SwitchToTab command's icon
-            SwitchToTabCommand().Icon(_lastIconPath);
+            // TODO CARLOS: We crash without this check :(
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalTab.cpp:192]: These two `TODO CARLOS` are part of why this is still in draft. `SwitchToTabCommand` is null for some reason. This is most likely related to the other blocking issue: the tabs have no content inside of them.

If I can get a fresh pair of eyes on this issue, this'll be ready for review/merge.

**Comment by DHowett** [comment]
_inheritance isn't really a thing people do in winrt_...

**Comment by DHowett** [comment]
If you're going to make it a base class, don't name it `I`Anything. `I` has a very specific meaning.

Is it possible to achieve what you're trying to do with a CRTP mix-in like IINheritable?

**Comment by zadjii-msft** [comment]
I haven't really reviewed this, but I'm doing something silly and similar [over here](https://github.com/microsoft/terminal/tree/dev/migrie/oop-monarch-peasant-sample/src/tools/MonarchPeasantSample)

In that branch, both the `Monarch` and `Peasant` projected types implement the `IPeasant` interface. On the implementation side, they both are derived from `PeasantBase`, which actually implements the entire `IPeasant` interface

**Comment by carlos-zamora** [comment]
For whatever reason, the following crash only occurs with a settings tab:

error message

<details>
<summary>Call Stack</summary>
```
>	Windows.UI.Xaml.dll!GetProgrammaticHitTestingParams(CCoreServices * coreServices, CUIElement * uiElement, ITransformer * * transformer, CUIElement * * ppThisNoRef, CPopupRoot * * popupRootNoRef, CUIElement * * popupHitTestSubtreeRootNoRef, bool * isRootNull) Line 5392	C++
 	Windows.UI.Xaml.dll!UIElement_HitTestPoint(CUIElement * uiElement, CCoreServices * coreServices, XPOINTF ptHit, bool canHitDisabledElements, bool canHitInvisibleElements, int * pnCount, CUIElement * * * ppElements) Line 5510	C++
 	Windows.UI.Xaml.dll!DirectUI::VisualTreeHelper::FindElementsInHostCoordinatesPointStatic(Windows::Foundation::Point intersectingPoint, Windows::UI::Xaml::IUIElement * pSubTree, unsigned char ppElements, unsigned char) Line 250	C++
 	Windows.UI.Xaml.dll!DirectUI::VisualTreeHelper::FindAllElementsInHostCoordinatesPointStatic(Windows::Foundation::Point intersectingPoint, Windows::UI::Xaml::IUIElement * pSubTree, unsigned char includeAllElements, Windows::Foundation::Collections::IIterable<Windows::UI::Xaml::UIElement *> * * ppElements) Line 329	C++
 	Windows.UI.Xaml.dll!DirectUI::CascadingMenuHelper::OnPointerExited(Windows::UI::Xaml::Input::IPointerRoutedEventArgs * args, bool parentIsSubMenu) Line 285	C++
 	Windows.UI.Xaml.dll!DirectUI::MenuFlyoutSubItem::OnPointerExited(Windows::UI::Xaml::Input::IPointerRoutedEventArgs * args) Line 141	C++
 	Windows.UI.Xaml.dll!DirectUI::ControlGenerated::OnPointerExitedProtected(Windows::UI::Xaml::Input::IPointerRoutedEventArgs * pE) Line 1404	C++
 	Windows.UI.Xaml.dll!DirectUI::Control::FireEvent(KnownEventIndex nDelegate, DirectUI::DependencyObject * pSender, IInspectable * pArgs) Line 267	C++
 	[Inline Frame] Windows.UI.Xaml.dll!DirectUI::DXamlCore::FireEvent(CDependencyObject *) Line 1985	C++
 	[Inline Frame] Windows.UI.Xaml.dll!AgCoreCallbacks::FireEvent(CDependencyObject *) Line 92	C++
 	[Inline Frame] Windows.UI.Xaml.dll!CFxCallbacks::JoltHelper_FireEvent(CDependencyObject *) Line 1015	C++
 	Windows.UI.Xaml.dll!CCoreServices::CLR_FireEvent(CDependencyObject * pListener, EventHandle hEvent, CDependencyObject * pSender, CEventArgs * pArgs, unsigned int flags) Line 3229	C++
 	Windows.UI.Xaml.dll!CommonBrowserHost::CLR_FireEvent(CDependencyObject * pListener, EventHandle hEvent, CDependencyObject * pSender, CEventArgs * pArgs, unsigned int flags) Line 771	C++
 	Windows.UI.Xaml.dll!CControlBase::ScriptCallback(void * pControl, CDependencyObject * pListener, EventHandle hEvent, CDependencyObject * pSender, CEventArgs * pArgs, int flags, IScriptObject * pScriptObject, HRESULT(*)(CDependencyObject *, CEventArgs *) pInternalHandler) Line 267	C++
 	Windows.UI.Xaml.dll!CXcpDispatcher::OnScriptCallback(CEventInfo * pEventInfo) Line 1347	C++
 	Windows.UI.Xaml.dll!CXcpDispatcher::OnWindowMessage(HWND__ * msg, unsigned int wParam, unsigned __int64 lParam, __int64) Line 1189	C++
 	Windows.UI.Xaml.dll!CXcpDispatcher::ProcessMessage(HWND__ * msg, unsigned int wParam, unsigned __int64 lParam, __int64 plRet, __int64 * pbDoDefault, unsigned int *) Line 925	C++
 	Windows.UI.Xaml.dll!CXcpDispatcher::WindowProc(HWND__ * hwnd, unsigned int msg, unsigned __int64 wParam, __int64 lParam) Line 839	C++
 	user32.dll!UserCallWinProcCheckWow(_ACTIVATION_CONTEXT * pActCtx, __int64(*)(tagWND *, unsigned int, unsigned __int64, __int64) pfn, HWND__ * hwnd, _WM_VALUE msg, unsigned __int64 wParam, __int64 lParam, void * fEnableLiteHooks, int) Line 280	C++
 	user32.dll!DispatchClientMessage(tagWND * pwnd, unsigned int message, unsigned __int64 wParam, __int64 lParam, unsigned __int64 pfn) Line 3446	C++
 	user32.dll!__fnDWORD(_FNDWORDMSG * pmsg) Line 1214	C++
 	ntdll.dll!KiUserCallbackDispatch() Line 603	Unknown
 	win32u.dll!ZwUserMessageCall() Line 213	Unknown
 	user32.dll!SendMessageWorker(tagWND * pwnd, unsigned int message, unsigned __int64 wParam, __int64 lParam, int fAnsi) Line 663	C++
 	user32.dll!SendMessageW(HWND__ * hwnd, unsigned int message, unsigned __int64 wParam, __int64 lParam) Line 905	C++
 	[Inline Frame] Windows.UI.Xaml.dll!CXcpDispatcher::SendMessageW(unsigned int) Line 809	C++
 	Windows.UI.Xaml.dll!CXcpBrowserHost::SyncScriptCallbackRequest(void * pVoidBH, CDependencyObject * pListener, EventHandle hEvent, CDependencyObject * pSender, CEventArgs * pArgs, int flags, IScriptObject * pScriptObject, HRESULT(*)(CDependencyObject *, CEventArgs *) pHandler) Line 1048	C++
 	[Inline Frame] Windows.UI.Xaml.dll!CEventManager::RaiseControlEvents(EventHandle) Line 1133	C++
 	Windows.UI.Xaml.dll!CEventManager::Raise(EventHandle hEvent, int bRefire, CDependencyObject * pSender, CEventArgs * pArgs, bool fRaiseSync, bool fInputEvent, bool bAllowErrorFallback, CDependencyObject * pSenderOverride) Line 887	C++
 	Windows.UI.Xaml.dll!ContentRootInput::PointerInputProcessor::ProcessPointerEnterLeave(CDependencyObject * pContactElement, CDependencyObject * pPointerEnterDO, unsigned int pointerId, CPointerEventArgs * pArgs, int bSkipLeave, unsigned int bForceRaisePointerEntered, unsigned int bIgnoreHitTestVisibleForPointerExited, unsigned int bAsyncEvent, unsigned int bAddEventRequest, bool * enterLeaveFound) Line 1446	C++
 	Windows.UI.Xaml.dll!ContentRootInput::PointerInputProcessor::ProcessPointerInput(InputMessage * pMsg, int * handled) Line 558	C++
 	Windows.UI.Xaml.dll!CInputServices::ProcessInput(InputMessage * pMsg, CContentRoot * contentRoot, int * handled) Line 888	C++
 	Windows.UI.Xaml.dll!CCoreServices::ProcessInput(InputMessage * pMessage, CContentRoot * contentRoot, int * fHandled) Line 992	C++
 	Windows.UI.Xaml.dll!CXcpBrowserHost::HandleInputMessage(unsigned int uMsg, MsgPacket * pMsgPack, CContentRoot * contentRoot, bool & fHandled) Line 1411	C++
 	Windows.UI.Xaml.dll!CJupiterControl::HandlePointerMessage(unsigned int uMsg, unsigned __int64 wParam, __int64 lParam, CContentRoot * contentRoot, Windows::UI::Input::IPointerPoint * pointerPoint, Windows::UI::Core::IPointerEventArgs * pointerArgs) Line 750	C++
 	Windows.UI.Xaml.dll!CJupiterControl::ForwardWindowedPopupMessageToJupiterWindow(HWND__ * window, unsigned int message, unsigned __int64 wParam, __int64 lParam, CContentRoot * contentRoot) Line 1191	C++
 	[Inline Frame] Windows.UI.Xaml.dll!DirectUI::DXamlCore::ForwardWindowedPopupMessageToJupiterWindow(HWND__ *) Line 4515	C++
 	Windows.UI.Xaml.dll!DirectUI::DXamlCore::ForwardWindowedPopupMessageToJupiterWindow(HWND__ * window, unsigned int message, unsigned __int64 wParam, __int64 lParam, CContentRoot * contentRoot, __int64 * pResult) Line 4482	C++
 	[Inline Frame] Windows.UI.Xaml.dll!CFxCallbacks::Core_ForwardWindowedPopupMessageToJupiterWindow(HWND__ * contentRoot, unsigned int) Line 1132	C++
 	Windows.UI.Xaml.dll!CPopup::HandleWindowedPopupMessage(HWND__ * hwnd, unsigned int message, unsigned __int64 wParam, __int64 lParam) Line 1652	C++
 	Windows.UI.Xaml.dll!CPopup::WindowedPopupWindowProc(HWND__ * hwnd, unsigned int message, unsigned __int64 wParam, __int64 lParam) Line 1625	C++
 	user32.dll!UserCallWinProcCheckWow(_ACTIVATION_CONTEXT * pActCtx, __int64(*)(tagWND *, unsigned int, unsigned __int64, __int64) pfn, HWND__ * hwnd, _WM_VALUE msg, unsigned __int64 wParam, __int64 lParam, void * fEnableLiteHooks, int) Line 280	C++
 	user32.dll!DispatchMessageWorker(tagMSG * pmsg, int fAnsi) Line 3157	C++
 	WindowsTerminal.exe!wWinMain(HINSTANCE__ * __formal, HINSTANCE__ * __formal, wchar_t * __formal, int __formal) Line 134	C++
 	[Inline Frame] WindowsTerminal.exe!invoke_main() Line 118	C++
 	WindowsTerminal.exe!__scrt_common_main_seh() Line 288	C++
 	kernel32.dll!BaseThreadInitThunk(unsigned long RunProcessInit, long(*)(void *) StartAddress, void * Argument) Line 70	C
 	ntdll.dll!RtlUserThreadStart(long(*)(void *) StartAddress, void * Argument) Line 1152	C
```
</details>

I get this by doing the following:
1. Open Settings UI tab
2. Open Close... submenu
3. Select "Close other tabs"
4. After the tabs close, right-click on the Settings UI tab again
5. Hover over Close... option

What's bizarre is that this _only_ happens with a Settings UI tab, not a Terminal Tab. But all of the submenu logic is in `TabBase`. Could I get some guidance here?

**Comment by carlos-zamora** [comment]
Good and bad news: the above bug seems to be a XAML Islands bug. It actually repros on the feature branch _before_ this change even goes in effect. So we're just gonna push forward here.

File: src/cascadia/TerminalApp/TabBase.cpp [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    WUX::FocusState TabBase::FocusState() const noexcept
+    {
+        return _focusState;
+    }
+
+    // Method Description:
+    // - Prepares this tab for being removed from the UI hierarchy
+    void TabBase::Shutdown()
+    {
+        // TODO: Does/Will the settings UI need some shutdown procedures?
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TabBase.cpp:33]: probably, but we're not equipped for that

**Review by DHowett — APPROVED** [review]
While you're here, can you delete the old `Tab.idl`?

@zadjii-msft Composition is bad for various reasons, and I don't fully understand it (it requires the creation of a wrapper class, and there's this inner/outer class divide, where the outer class (TabBase) actually delegates to the inner class or something?)... but... doing the `ConnectionStateHolder<T>` or `IInheritable<T>` thing is even _worse_ because we'd need to template it (again) so that we could dispatch the events with the right type[1].

[1]: we can't dispatch typed events off a TabBase, because the typed event's first argument is an IInspectable. TabBase would be _outside the hierarchy_, and not convertible to IInspectable. It's a whole ordeal to get it wired up, and it would require even more new macros like `OBSERVABLE_GETSET_MACRO_BUT_WITH_EVENT_TYPE_T`.

**Comment by zadjii-msft** [comment]
> It actually repros on the feature branch _before_ this change even goes in effect.

That raises even MORE questions❗ We should probably consider that a blocking bug for the SUI, right? Since it's pretty trivially reproable, and the severity is _the whole app crashes_

**Review by zadjii-msft — APPROVED** [review]
I'm very concerned about the crash in `feature/settings-ui`, but this is okay

**Pull request merged**